Utils/azuretools-core/src/com/microsoft/azuretools/adauth/AuthContext.java

```diff
@@ -128,6 +135,73 @@ public AuthResult acquireToken(@NotNull String resource, boolean newAuthCode,
         }
     }
 
+    /**
+     * Acquire token for resource by device code.
+     */
+    public AuthResult acquireToken(@NotNull final String resource, final boolean newDeviceCode, final String uid,
+                                   final AuthenticationCallback<AuthenticationResult> callback) throws AuthException {
+        try {
+            if (!newDeviceCode) {
+                final AuthResult result = acquireTokenFromCache(resource, uid);
+                if (result != null) {
+                    return result;
+                } else {
+                    throw new AuthException(AuthError.UnknownUser);
+                }
+            }
+            final ExecutorService service = Executors.newFixedThreadPool(5);
+            final AuthenticationContext ctx = new AuthenticationContext(authority, true, service);
+            final IDeviceLoginUI deviceLoginUI = CommonSettings.getUiFactory().getDeviceLoginUI();
+            final DeviceCode deviceCode = ctx.acquireDeviceCode(clientId, resource, null).get();
+            deviceLoginUI.showDeviceLoginMessage(deviceCode.getMessage());
+            final AuthResult result = acquireTokenByDeviceCode(ctx, resource, deviceCode, deviceLoginUI, callback);
+            deviceLoginUI.close();
```

Comment:
How about to put `close()` invoking into the Device Code login modal dialog? The result above can be `null` if the user cancel the dialog.